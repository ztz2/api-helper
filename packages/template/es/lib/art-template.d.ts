import artTemplate from './utils/template/art-template-web.js';
export default artTemplate;
