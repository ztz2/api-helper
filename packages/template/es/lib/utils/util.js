var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
import { filterSchemaPrimitiveValue } from '@api-helper/core/lib/utils/util';
import formatCode from '../../lib/utils/prettier';
export function postCode(_a, _b) {
    var _c = _a.code, code = _c === void 0 ? '' : _c, _d = _a.ki, ki = _d === void 0 ? '' : _d, _e = _a.commentCode, commentCode = _e === void 0 ? '' : _e;
    var _f = _b.onlyBody, onlyBody = _f === void 0 ? false : _f;
    // 移除多个空格
    ki = ki.replace(/\s(\s+)/gim, ' ');
    // 移除双换行符。
    code = code.replace(/\n\n/gim, '\n');
    // 格式化代码
    code = formatCode(ki + code, {
        parser: 'typescript'
    });
    // 如果只输出body部分，移除 export xxx 部分
    if (onlyBody) {
        code = code.replace(ki, '');
    }
    // 合并注释代码输出
    return [commentCode, code].filter(Boolean).join('\n');
}
export function checkIsInterface(schema) {
    var _a;
    var isType = !schema || (schema === null || schema === void 0 ? void 0 : schema.type) === 'array' || isEmptyObject(schema);
    var hasKey = (_a = schema === null || schema === void 0 ? void 0 : schema.params) === null || _a === void 0 ? void 0 : _a.some(function (s) { return s.keyName; });
    return !isType && hasKey;
}
export function isEmptyObject(schema) {
    var _a;
    return (schema === null || schema === void 0 ? void 0 : schema.type) === 'object' && ((_a = schema === null || schema === void 0 ? void 0 : schema.params) === null || _a === void 0 ? void 0 : _a.length) === 0;
}
export function isEmptySchema(schema) {
    var exec = function (scm) {
        var e_1, _a;
        if (!scm) {
            return true;
        }
        if (scm.type === 'object' && isEmptyObject(filterSchemaPrimitiveValue(scm))) {
            return true;
        }
        if (scm.type === 'array') {
            try {
                for (var _b = __values(scm.params), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var itm = _c.value;
                    if (exec(itm)) {
                        return true;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return false;
        }
        return false;
    };
    return exec(schema);
}
