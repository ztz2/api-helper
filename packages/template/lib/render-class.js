"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderClassName = exports.renderClass = void 0;
var merge_1 = __importDefault(require("lodash/merge"));
var cloneDeep_1 = __importDefault(require("lodash/cloneDeep"));
var _changeCase = __importStar(require("change-case"));
var util_1 = require("@api-helper/core/lib/utils/util");
var util_2 = require("../lib/utils/util");
var render_object_1 = require("../lib/render-object");
function renderClass(schema, api, options) {
    var _a;
    options = (0, merge_1.default)({
        onlyBody: false,
        prefix: 'export ',
        paramType: 'request',
        emptyBodyCode: '{}',
        suffixName: 'class',
    }, options);
    schema = (0, cloneDeep_1.default)(schema);
    schema = (0, render_object_1.precessArraySchema)(schema);
    // const schemaList = filterSchemaPrimitiveValue(Array.isArray(schema) ? schema : (schema as APIHelper.Schema)?.params ?? []) as APIHelper.SchemaList;
    var primitiveValueSchema = (0, util_1.filterSchemaPrimitiveValue)(schema);
    var prefix = options.prefix, suffixName = options.suffixName, onlyBody = options.onlyBody, dropComment = options.dropComment, emptyBodyCode = options.emptyBodyCode, _b = options.paramType, paramType = _b === void 0 ? 'request' : _b;
    var keyword = "".concat(prefix, " class");
    var onRenderClassName = (options === null || options === void 0 ? void 0 : options.onRenderClassName) ? options.onRenderClassName : renderClassName;
    var commentCode = onlyBody ? '' : dropComment !== true ? renderClassComment(api, paramType) : '';
    var className = (options === null || options === void 0 ? void 0 : options.name) ? options.name : onRenderClassName(api, {
        paramType: paramType,
        suffixName: suffixName,
        changeCase: _changeCase,
    });
    /**
     * output ->  export class
     */
    var ki = ["".concat(keyword, " ").concat(className, " ")].filter(Boolean).join('\n');
    var bodyCode = '';
    if (!schema || (((_a = schema === null || schema === void 0 ? void 0 : schema.params) === null || _a === void 0 ? void 0 : _a.length) === 0 && !(schema === null || schema === void 0 ? void 0 : schema.keyName)) || ((schema === null || schema === void 0 ? void 0 : schema.type) !== 'object' && (schema === null || schema === void 0 ? void 0 : schema.type) !== 'array')) {
        if ((schema === null || schema === void 0 ? void 0 : schema.type) !== 'object' && (schema === null || schema === void 0 ? void 0 : schema.type) !== 'array') {
            bodyCode += '/* 非对象数据，不能生成Class代码**/\n';
        }
        bodyCode += emptyBodyCode;
    }
    else {
        bodyCode = renderClassDeepObject(primitiveValueSchema, true);
    }
    return (0, util_2.postCode)({
        ki: ki,
        commentCode: commentCode,
        code: bodyCode,
    }, { onlyBody: onlyBody });
}
exports.renderClass = renderClass;
function renderClassDeepObject(schema, isRoot, memo) {
    var e_1, _a;
    var _b;
    if (isRoot === void 0) { isRoot = false; }
    if (memo === void 0) { memo = new Map(); }
    if (memo.has(schema)) {
        return memo.get(schema);
    }
    memo.set(schema, 'null');
    var codeWrap = [];
    var prefix = (schema === null || schema === void 0 ? void 0 : schema.type) === 'array' ? '[\n' : '{\n';
    var postfix = (schema === null || schema === void 0 ? void 0 : schema.type) === 'array' ? '\n]' : '\n}';
    try {
        for (var _c = __values(schema.params), _d = _c.next(); !_d.done; _d = _c.next()) {
            var child = _d.value;
            var keyName = (_b = child.keyName) !== null && _b !== void 0 ? _b : '';
            var type = child.type;
            var temporaryCode = [(0, render_object_1.renderComment)(child)];
            var evaluationCode = isRoot ? ' = ' : ': ';
            var v = "''";
            switch (type) {
                // 数组类型 | 对象类型
                case 'array':
                case 'object':
                    temporaryCode.pop();
                    temporaryCode.push(renderClassDeepObject(child, false, memo));
                    break;
                case 'string':
                    if ('enum' in child && child.enum.length > 0) {
                        v = "'".concat(child.enum[0], "'");
                    }
                    temporaryCode.push("".concat((0, util_1.processKeyName)(keyName)).concat(evaluationCode).concat(v));
                    break;
                case 'number':
                    temporaryCode.push("".concat((0, util_1.processKeyName)(keyName)).concat(evaluationCode, "0"));
                    break;
                case 'boolean':
                    temporaryCode.push("".concat((0, util_1.processKeyName)(keyName)).concat(evaluationCode, "false"));
                    break;
                // 其他类型
                default:
                    temporaryCode.push("".concat((0, util_1.processKeyName)(keyName)).concat(evaluationCode, "null"));
            }
            codeWrap.push(temporaryCode.join('\n'));
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
        }
        finally { if (e_1) throw e_1.error; }
    }
    var code = prefix + codeWrap.join(isRoot ? ';\n' : ',\n') + postfix;
    if (schema === null || schema === void 0 ? void 0 : schema.keyName) {
        var evaluationCode = isRoot ? ' = ' : ': ';
        code = [
            (0, render_object_1.renderComment)(schema),
            "".concat((0, util_1.processKeyName)(schema.keyName)).concat(evaluationCode, " ").concat(code)
        ].join('\n');
        if (isRoot && schema.type !== 'object') {
            code = "{ \n ".concat(code, " \n } \n");
        }
    }
    return code;
}
function renderClassName(api, options) {
    var _a, _b;
    var suffixName = (_a = options === null || options === void 0 ? void 0 : options.suffixName) !== null && _a !== void 0 ? _a : 'Class ';
    var changeCase = (_b = options.changeCase) !== null && _b !== void 0 ? _b : _changeCase;
    var name = api.path;
    if (options.paramType) {
        name += " ".concat(options.paramType);
    }
    name += "".concat(suffixName, "By ").concat(api.method);
    return changeCase.pascalCase(name);
}
exports.renderClassName = renderClassName;
function renderClassComment(api, paramType, isExtraData) {
    if (isExtraData === void 0) { isExtraData = false; }
    var summary = isExtraData ? 'Incompatible request data' : paramType === 'request' ? 'Request data' : paramType === 'response' ? 'Response data' : '';
    if (summary) {
        summary += ' entity class';
    }
    var commentText = "/**\n * @description ".concat([api.title, api.description].filter(Boolean).join('、')).concat(summary ? "\n * @summary ".concat(summary) : '').concat(api.docURL ? "\n * @doc ".concat(api.docURL) : '', "\n * @url [ ").concat(api.method.toUpperCase(), " ] ").concat(api.path, "\n */");
    return commentText;
}
