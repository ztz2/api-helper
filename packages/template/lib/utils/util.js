"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isEmptyObject = exports.checkIsInterface = exports.postCode = void 0;
var prettier_1 = __importDefault(require("../../lib/utils/prettier"));
function postCode(_a, _b) {
    var _c = _a.code, code = _c === void 0 ? '' : _c, _d = _a.ki, ki = _d === void 0 ? '' : _d, _e = _a.commentCode, commentCode = _e === void 0 ? '' : _e;
    var _f = _b.onlyBody, onlyBody = _f === void 0 ? false : _f;
    // 移除多个空格
    ki = ki.replace(/\s(\s+)/gim, ' ');
    // 移除双换行符。
    code = code.replace(/\n\n/gim, '\n');
    // 格式化代码
    code = (0, prettier_1.default)(ki + code, {
        parser: 'typescript'
    });
    // 如果只输出body部分，移除 export xxx 部分
    if (onlyBody) {
        code = code.replace(ki, '');
    }
    // 合并注释代码输出
    return [commentCode, code].filter(Boolean).join('\n');
}
exports.postCode = postCode;
function checkIsInterface(schema) {
    var _a;
    var isType = !schema || (schema === null || schema === void 0 ? void 0 : schema.type) === 'array' || isEmptyObject(schema);
    var hasKey = (_a = schema === null || schema === void 0 ? void 0 : schema.params) === null || _a === void 0 ? void 0 : _a.some(function (s) { return s.keyName; });
    return !isType && hasKey;
}
exports.checkIsInterface = checkIsInterface;
function isEmptyObject(schema) {
    var _a;
    return (schema === null || schema === void 0 ? void 0 : schema.type) === 'object' && ((_a = schema === null || schema === void 0 ? void 0 : schema.params) === null || _a === void 0 ? void 0 : _a.length) === 0;
}
exports.isEmptyObject = isEmptyObject;
