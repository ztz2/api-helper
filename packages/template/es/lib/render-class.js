var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
import merge from 'lodash/merge';
import cloneDeep from 'lodash/cloneDeep';
import * as changeCase from 'change-case';
import { filterSchemaPrimitiveValue } from '@api-helper/core/lib/utils/util';
import { postCode } from '../lib/utils/util';
import { renderComment } from '../lib/render-object';
export function renderClass(schema, api, options) {
    var _a, _b;
    options = merge({
        onlyBody: false,
        prefix: 'export ',
        paramType: 'request',
        emptyBodyCode: '{}',
    }, options);
    schema = cloneDeep(schema);
    var schemaList = filterSchemaPrimitiveValue(Array.isArray(schema) ? schema : (_b = (_a = schema) === null || _a === void 0 ? void 0 : _a.params) !== null && _b !== void 0 ? _b : []);
    var prefix = options.prefix, onlyBody = options.onlyBody, dropComment = options.dropComment, emptyBodyCode = options.emptyBodyCode, _c = options.paramType, paramType = _c === void 0 ? 'request' : _c;
    var keyword = "".concat(prefix, " class");
    var onRenderClassName = (options === null || options === void 0 ? void 0 : options.onRenderClassName) ? options.onRenderClassName : renderClassName;
    var commentCode = onlyBody ? '' : dropComment !== true ? renderClassComment(api, paramType) : '';
    var className = (options === null || options === void 0 ? void 0 : options.name) ? options.name : onRenderClassName(api, {
        paramType: paramType,
        changeCase: changeCase
    });
    /**
     * output ->  export class
     */
    var ki = ["".concat(keyword, " ").concat(className, " ")].filter(Boolean).join('\n');
    var bodyCode;
    if (!schema || (Array.isArray(schema) && schema.length === 0)) {
        bodyCode = emptyBodyCode;
    }
    else {
        bodyCode = renderClassDeepObject(schemaList, null, true);
    }
    return postCode({
        ki: ki,
        commentCode: commentCode,
        code: bodyCode,
    }, { onlyBody: onlyBody });
}
function renderClassDeepObject(schemaList, parentSchema, isRoot, memo) {
    var e_1, _a;
    var _b;
    if (parentSchema === void 0) { parentSchema = null; }
    if (isRoot === void 0) { isRoot = false; }
    if (memo === void 0) { memo = new Map(); }
    if (memo.has(schemaList)) {
        return memo.get(schemaList);
    }
    memo.set(schemaList, 'null');
    var codeWrap = [];
    var prefix = (parentSchema === null || parentSchema === void 0 ? void 0 : parentSchema.type) === 'array' ? '[\n' : '{\n';
    var postfix = (parentSchema === null || parentSchema === void 0 ? void 0 : parentSchema.type) === 'array' ? '\n]' : '\n}';
    try {
        for (var schemaList_1 = __values(schemaList), schemaList_1_1 = schemaList_1.next(); !schemaList_1_1.done; schemaList_1_1 = schemaList_1.next()) {
            var schema = schemaList_1_1.value;
            var keyName = (_b = schema.keyName) !== null && _b !== void 0 ? _b : '';
            var type = schema.type;
            var temporaryCode = [renderComment(schema)];
            var evaluationCode = isRoot ? ' = ' : ': ';
            var v = "''";
            switch (type) {
                // 数组类型 | 对象类型
                case 'array':
                case 'object':
                    temporaryCode.pop();
                    temporaryCode.push(renderClassDeepObject(schema.params, schema));
                    break;
                case 'string':
                    if ('enum' in schema && schema.enum.length > 0) {
                        v = "'".concat(schema.enum[0], "'");
                    }
                    temporaryCode.push("".concat(keyName).concat(evaluationCode).concat(v));
                    break;
                case 'number':
                    temporaryCode.push("".concat(keyName).concat(evaluationCode, "0"));
                    break;
                case 'boolean':
                    temporaryCode.push("".concat(keyName).concat(evaluationCode, "false"));
                    break;
                // 其他类型
                default:
                    temporaryCode.push("".concat(keyName).concat(evaluationCode, "null"));
            }
            codeWrap.push(temporaryCode.join('\n'));
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (schemaList_1_1 && !schemaList_1_1.done && (_a = schemaList_1.return)) _a.call(schemaList_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
    var code = prefix + codeWrap.join(isRoot ? ';\n' : ',\n') + postfix;
    if (parentSchema === null || parentSchema === void 0 ? void 0 : parentSchema.keyName) {
        code = [
            renderComment(parentSchema),
            "".concat(parentSchema.keyName, ": ").concat(code)
        ].join('\n');
    }
    return code;
}
export function renderClassName(api, options) {
    var name = api.path;
    if (options.paramType) {
        name += " ".concat(options.paramType);
    }
    name += "By ".concat(api.method);
    return changeCase.pascalCase(name);
}
function renderClassComment(api, paramType, isExtraData) {
    if (isExtraData === void 0) { isExtraData = false; }
    var commentText = "/**\n * @description ".concat([api.title, api.description].filter(Boolean).join('、'), "\u3010").concat(isExtraData ? '不兼容的请求数据' : paramType === 'request' ? '请求数据' : paramType === 'response' ? '响应数据' : '', "\u5B9E\u4F53\u7C7B\u3011").concat(api.docURL ? "\n * @doc ".concat(api.docURL) : '', "\n * @url [ ").concat(api.method.toUpperCase(), " ] ").concat(api.path, "\n */");
    return commentText;
}
