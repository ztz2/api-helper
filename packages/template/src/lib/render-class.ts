import merge from 'lodash/merge';
import cloneDeep from 'lodash/cloneDeep';
import * as _changeCase from 'change-case';
import { APIHelper } from '@api-helper/core/lib/types';
import { filterSchemaPrimitiveValue } from '@api-helper/core/lib/utils/util';

import { ChangeCase } from '@/lib/types';
import { postCode } from '@/lib/utils/util';
import { renderComment } from '@/lib/render-object';

export function renderClass(
  schema: APIHelper.Schema | Array<APIHelper.Schema> | null,
  api: APIHelper.API,
  options?: {
    // 请求数据还是响应数据
    paramType?: 'request' | 'response';
    // 前缀
    prefix?: string,
    // interface 名称
    name?: string;
    // 删除interface注释
    dropComment?: boolean;
    // 只生成类型部分
    onlyBody?: boolean;
    // 当数据为空的body代码
    emptyBodyCode?: string;
    // 生成接口名称回调
    onRenderClassName?: typeof renderClassName;
  }
) {
  options = merge({
    onlyBody: false,
    prefix: 'export ',
    paramType: 'request',
    emptyBodyCode: '{}',
  }, options);

  schema = cloneDeep(schema);

  const schemaList = filterSchemaPrimitiveValue(Array.isArray(schema) ? schema : (schema as APIHelper.Schema)?.params ?? []) as APIHelper.SchemaList;

  const {
    prefix,
    onlyBody,
    dropComment,
    emptyBodyCode,
    paramType = 'request',
  } = options;
  const keyword = `${prefix} class`;
  const onRenderClassName = options?.onRenderClassName ? options.onRenderClassName : renderClassName;
  let commentCode = onlyBody ? '' : dropComment !== true ? renderClassComment(api, paramType) : '';
  let className = options?.name ? options.name : onRenderClassName(api, {
    paramType,
    changeCase: _changeCase,
  });

  /**
   * output ->  export class
   */
  let ki = [`${keyword} ${className} `].filter(Boolean).join('\n');

  let bodyCode;
  if (!schema || (Array.isArray(schema) && schema.length === 0)){
    bodyCode = emptyBodyCode;
  } else {
    bodyCode = renderClassDeepObject(schemaList, null, true);
  }

  return postCode({
    ki,
    commentCode,
    code: bodyCode,
  }, { onlyBody });
}

function renderClassDeepObject(
  schemaList: Array<APIHelper.Schema>,
  parentSchema: APIHelper.Schema | null = null,
  isRoot = false,
  memo = new Map<APIHelper.Schema[], string>()
): string {
  if (memo.has(schemaList)) {
    return memo.get(schemaList) as string;
  }

  memo.set(schemaList, 'null');

  const codeWrap: string[] = [];
  const prefix = parentSchema?.type === 'array' ? '[\n' : '{\n';
  const postfix = parentSchema?.type === 'array' ? '\n]' : '\n}';
  for (const schema of schemaList) {
    const keyName = schema.keyName ?? '';

    const type = schema.type;
    const temporaryCode = [renderComment(schema)];
    const evaluationCode = isRoot ? ' = ' : ': ';
    let v = "''";
    switch (type) {
      // 数组类型 | 对象类型
      case 'array': case 'object':
        temporaryCode.pop();
        temporaryCode.push(renderClassDeepObject(schema.params, schema));
        break;
      case 'string':
        if ('enum' in schema && schema.enum.length > 0) {
          v = `'${schema.enum[0]}'`;
        }
        temporaryCode.push(`${keyName}${evaluationCode}${v}`);
        break;
      case 'number':
        temporaryCode.push(`${keyName}${evaluationCode}0`);
        break;
      case 'boolean':
        temporaryCode.push(`${keyName}${evaluationCode}false`);
        break;
      // 其他类型
      default:
        temporaryCode.push(`${keyName}${evaluationCode}null`);
    }
    codeWrap.push(temporaryCode.join('\n'));
  }

  let code = prefix + codeWrap.join(isRoot ? ';\n' : ',\n') + postfix;

  if (parentSchema?.keyName) {
    code = [
      renderComment(parentSchema),
      `${parentSchema.keyName}: ${code}`
    ].join('\n');
  }

  return code;
}

export function renderClassName(
  api: APIHelper.API,
  options: {
    paramType: 'request' | 'response';
    changeCase: ChangeCase;
  }) {
  const changeCase = options.changeCase ?? _changeCase;
  let name = api.path;
  if (options.paramType) {
    name += ` ${options.paramType}`;
  }
  name += `By ${api.method}`;
  return changeCase.pascalCase(name);
}

function renderClassComment(
  api: APIHelper.API,
  paramType: 'request' | 'response',
  isExtraData = false
) {
  const commentText =
    `/**
 * @description ${[api.title, api.description].filter(Boolean).join('、')}【${isExtraData ? '不兼容的请求数据' : paramType === 'request' ? '请求数据' : paramType === 'response' ? '响应数据' : ''}实体类】${api.docURL ? `\n * @doc ${api.docURL}` : ''}
 * @url [ ${api.method.toUpperCase()} ] ${api.path}
 */`;
  return commentText;
}
