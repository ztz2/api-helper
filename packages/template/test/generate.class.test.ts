import { join } from 'path';
import { APIHelper } from '@api-helper/core';
import ParserSwagger from '@api-helper/core/es/lib/parser/parser-swagger';
import { filterSchemaPrimitiveValue } from '@api-helper/core/lib/utils/util';

import { readJsonFile } from './utils/file';
import { renderClass } from '../src/lib/render-class';

describe('生成 Class 类测试', () => {
  test('OpenAPI-2.0生成 Class 类测试', async () => {
    const openAPIDocumentJSON = await readJsonFile(join(__dirname, '../../core/test/resources/open-api/v2.0/petstore.json'));
    const openAPIDocumentList = await new ParserSwagger(false).parser([openAPIDocumentJSON]);
    const openAPIDocument = openAPIDocumentList[0];
    const category = openAPIDocument.categoryList[1];
    const api = category.apiList[0];
    const code = renderClass(api.responseDataSchema, api, {
      paramType: 'request'
    });
    expect(
      code,
    ).toMatchSnapshot('OpenAPI-2.0生成 Class 类测试');
  });

  test('OpenAPI-3.0生成 Class 类测试', async () => {
    const openAPIDocumentJSON = await readJsonFile(join(__dirname, '../../core/test/resources/open-api/v3.0/petstore.json'));
    const openAPIDocumentList = await new ParserSwagger(false).parser([openAPIDocumentJSON]);
    const openAPIDocument = openAPIDocumentList[0];
    const category = openAPIDocument.categoryList[1];
    const api = category.apiList[0];
    const code = renderClass(api.requestDataSchema, api, {
      paramType: 'response'
    });
    expect(
      code,
    ).toMatchSnapshot('OpenAPI-3.00生成 Class 类测试');
  });
  test('name - prefix - dropComment - onlyBody 属性测试', async () => {
    const openAPIDocumentJSON = await readJsonFile(join(__dirname, '../../core/test/resources/open-api/v2.0/petstore.json'));
    const openAPIDocumentList = await new ParserSwagger(false).parser([openAPIDocumentJSON]);
    const openAPIDocument = openAPIDocumentList[0];
    const category = openAPIDocument.categoryList[1];
    const api = category.apiList[0];
    const schema = filterSchemaPrimitiveValue(api.responseDataSchema) as APIHelper.Schema;
    expect(
      renderClass(schema, api, {
        paramType: 'request',
        name: '自定义名称',
      }),
    ).toMatchSnapshot('OpenAPI-2.0生成 Class name属性测试');
    expect(
      renderClass(schema, api, {
        paramType: 'request',
        prefix: 'export',
      }),
    ).toMatchSnapshot('OpenAPI-2.0生成 Class prefix属性测试');
    expect(
      renderClass(schema, api, {
        paramType: 'request',
        dropComment: true,
      }),
    ).toMatchSnapshot('OpenAPI-2.0生成 Class dropComment属性测试');
    expect(
      renderClass(schema, api, {
        paramType: 'request',
        onlyBody: true,
      }),
    ).toMatchSnapshot('OpenAPI-2.0生成 Class onlyBody属性测试');
  });

  test('array单参数', async () => {
    const openAPIDocumentJSON = await readJsonFile(join(__dirname, '../../core/test/resources/open-api/v2.0/petstore.json'));
    const openAPIDocumentList = await new ParserSwagger(false).parser([openAPIDocumentJSON]);
    const openAPIDocument = openAPIDocumentList[0];
    const category = openAPIDocument.categoryList[0];
    const api = category.apiList[0];
    expect(
      renderClass({
        'id': 'bc2f818f91624c8c',
        'type': 'array',
        'keyName': 'storeIds',
        'title': '',
        'description': '',
        'label': '',
        'rules': { 'required': true },
        'examples': [],
        'params': [],
        'enum': [],
      }, api),
    ).toMatchSnapshot('array单参数');
  });
});
