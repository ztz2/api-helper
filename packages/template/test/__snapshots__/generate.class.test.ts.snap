// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`生成 Class 类测试 OpenAPI-2.0生成 Class 类测试: OpenAPI-2.0生成 Class 类测试 1`] = `
"/**
 * @description Place an order for a pet
 * @summary Request data entity class
 * @url [ POST ] /v2/store/order
 */
export class V2StoreOrderRequestclassByPost {
  // { string }
  id = '';
  // { string }
  petId = '';
  // { number }
  quantity = 0;
  // { string }
  shipDate = '';
  // { 'placed' | 'approved' | 'delivered' } Order Status
  status = 'placed';
  // { array } Order Tag
  tag: [
    [
      {
        // { 'error' | 'waring' | 'info' }
        tagName: 'error';
      }
    ]
  ];
  // { boolean }
  complete = false;
}
"
`;

exports[`生成 Class 类测试 OpenAPI-3.0生成 Class 类测试: OpenAPI-3.00生成 Class 类测试 1`] = `
"/**
 * @description Place an order for a pet、Place a new order in the store
 * @summary Response data entity class
 * @url [ POST ] /api/v3/store/order
 */
export class ApiV3StoreOrderResponseclassByPost {
  // { string }
  id = '';
  // { string }
  petId = '';
  // { number }
  quantity = 0;
  // { string }
  shipDate = '';
  // { 'placed' | 'approved' | 'delivered' } Order Status
  status = 'placed';
  // { boolean }
  complete = false;
}
"
`;

exports[`生成 Class 类测试 array单参数: array单参数 1`] = `
"/**
 * @description list
 * @summary Request data entity class
 * @url [ GET ] /system/user/list
 */
export class SystemUserListRequestclassByGet {
  // { array } 门店id
  storeIds = [];
}
"
`;

exports[`生成 Class 类测试 array单参数2: array单参数2 1`] = `
"/**
 * @description list
 * @summary Request data entity class
 * @url [ GET ] /system/user/list
 */
export class SystemUserListRequestclassByGet {
  // { array }
  storeIds = [];
}
"
`;

exports[`生成 Class 类测试 name - prefix - dropComment - onlyBody 属性测试: OpenAPI-2.0生成 Class dropComment属性测试 1`] = `
"export class V2StoreOrderRequestclassByPost {
  // { string }
  id = '';
  // { string }
  petId = '';
  // { number }
  quantity = 0;
  // { string }
  shipDate = '';
  // { 'placed' | 'approved' | 'delivered' } Order Status
  status = 'placed';
  // { array } Order Tag
  tag: [
    [
      {
        // { 'error' | 'waring' | 'info' }
        tagName: 'error';
      }
    ]
  ];
  // { boolean }
  complete = false;
}
"
`;

exports[`生成 Class 类测试 name - prefix - dropComment - onlyBody 属性测试: OpenAPI-2.0生成 Class name属性测试 1`] = `
"/**
 * @description Place an order for a pet
 * @summary Request data entity class
 * @url [ POST ] /v2/store/order
 */
export class 自定义名称 {
  // { string }
  id = '';
  // { string }
  petId = '';
  // { number }
  quantity = 0;
  // { string }
  shipDate = '';
  // { 'placed' | 'approved' | 'delivered' } Order Status
  status = 'placed';
  // { array } Order Tag
  tag: [
    [
      {
        // { 'error' | 'waring' | 'info' }
        tagName: 'error';
      }
    ]
  ];
  // { boolean }
  complete = false;
}
"
`;

exports[`生成 Class 类测试 name - prefix - dropComment - onlyBody 属性测试: OpenAPI-2.0生成 Class onlyBody属性测试 1`] = `
"{
  // { string }
  id = '';
  // { string }
  petId = '';
  // { number }
  quantity = 0;
  // { string }
  shipDate = '';
  // { 'placed' | 'approved' | 'delivered' } Order Status
  status = 'placed';
  // { array } Order Tag
  tag: [
    [
      {
        // { 'error' | 'waring' | 'info' }
        tagName: 'error';
      }
    ]
  ];
  // { boolean }
  complete = false;
}"
`;

exports[`生成 Class 类测试 name - prefix - dropComment - onlyBody 属性测试: OpenAPI-2.0生成 Class prefix属性测试 1`] = `
"/**
 * @description Place an order for a pet
 * @summary Request data entity class
 * @url [ POST ] /v2/store/order
 */
export class V2StoreOrderRequestclassByPost {
  // { string }
  id = '';
  // { string }
  petId = '';
  // { number }
  quantity = 0;
  // { string }
  shipDate = '';
  // { 'placed' | 'approved' | 'delivered' } Order Status
  status = 'placed';
  // { array } Order Tag
  tag: [
    [
      {
        // { 'error' | 'waring' | 'info' }
        tagName: 'error';
      }
    ]
  ];
  // { boolean }
  complete = false;
}
"
`;

exports[`生成 Class 类测试 空属性: 空属性 1`] = `
"/**
 * @description Place an order for a pet
 * @summary Request data entity class
 * @url [ POST ] /v2/store/order
 */
export class V2StoreOrderRequestclassByPost {
  /* 非对象数据，不能生成Class代码**/
}
"
`;

exports[`生成 Class 类测试 空属性: 空数组 1`] = `
"/**
 * @description Place an order for a pet
 * @summary Request data entity class
 * @url [ POST ] /v2/store/order
 */
export class V2StoreOrderRequestclassByPost {}
"
`;
